src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,298 @@
+  // mov rsi, rbx                     48 8b f3
+  // movabs rax, address_dlopen       48 b8 address_dlopen
+  // call rax                         ff d0
+  // int3                             cc
+  MachineCode code;
+  code.AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(address_so_path)
+      .AppendBytes({0x48, 0xc7, 0xc3})
+      .AppendImmediate32(flag)
+      .AppendBytes({0x48, 0x8b, 0xf8})
+      .AppendBytes({0x48, 0x8b, 0xf3})
+      .AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(address_dlopen)
+      .AppendBytes({0xff, 0xd0})
+      .AppendBytes({0xcc});
+  auto result_write_code = WriteTraceesMemory(pid, address_code, code.GetResultAsVector());
+  if (result_write_code.has_error()) {
+    FreeMemoryOrDie(pid, address_code, memory_size);
+    return result_write_code.error();
+  }
+
+  ExecuteOrDie(pid, original_registers, address_code);
+
+  void* return_value = absl::bit_cast<void*>(GetReturnValueOrDie(pid));
+
+  // Cleanup memory and registers.
+  RestoreRegistersOrDie(original_registers);
+  FreeMemoryOrDie(pid, address_code, memory_size);
+  return return_value;
+}
+
+[[nodiscard]] ErrorMessageOr<void*> DlsymInTracee(pid_t pid, void* handle, std::string symbol) {
+  // Figure out address of dlsym in libc.
+  OUTCOME_TRY(address_dlsym, FindFunctionAddress(pid, "dlsym", "libc-"));
+
+  // Backup registers.
+  RegisterState original_registers;
+  OUTCOME_TRY(original_registers.BackupRegisters(pid));
+
+  // Allocate small memory area in the tracee. This is used for the code and the symbol name.
+  const size_t symbol_name_length = symbol.length() + 1;  // include terminating zero
+  const uint64_t memory_size = kScratchPadSize + symbol_name_length;
+  OUTCOME_TRY(address_code, AllocateInTracee(pid, memory_size));
+
+  // Write the name of symbol into memory at address_code with offset of kScratchPadSize.
+  std::vector<uint8_t> symbol_name_as_vector(symbol_name_length);
+  memcpy(symbol_name_as_vector.data(), symbol.c_str(), symbol_name_length);
+  const uint64_t address_symbol_name = address_code + kScratchPadSize;
+  auto result_write_symbol_name =
+      WriteTraceesMemory(pid, address_symbol_name, symbol_name_as_vector);
+  if (result_write_symbol_name.has_error()) {
+    FreeMemoryOrDie(pid, address_code, memory_size);
+    return result_write_symbol_name.error();
+  }
+
+  // We want to do the following in the tracee:
+  // return_value = dlsym(handle, symbol);
+  // The calling convention is to put the parameters in registers rdi and rsi.
+  // So the handle goes to rdi and the address of the symbol name goes to rsi. Then we load the
+  // address of dlsym into rax and do the call. Assembly in Intel syntax (destination first),
+  // machine code on the right:
+
+  // movabs rax, handle               48 b8 handle
+  // mov rdi, rax                     48 89 c7
+  // movabs rax, address_symbol_name  48 b8 address_symbol_name
+  // mov rsi, rax                     48 89 c6
```

Comment:
Same question as above: can't you write `rdi` and `rsi` directly instead of going through `rax`?
Also, the pattern here is different than `DlopenInTracee`, as you reuse `rax` here, while above you use `rax` and `rbx`. I would try to be consistent.

Reply:
I think you've answered this.